### PR TITLE
Improve `metadata` recovery against EOF token

### LINKED ISSUE #29730: Improve `metadata` recovery at EOF
**Description:**
Currently, we invalidate `AnnotationNode` at `EOF` which causes problems for LS. Ref: Issue #29420

We could rather attach it to a dummy node. 
Also, we need to mark the node as a missing node as per https://github.com/ballerina-platform/ballerina-lang/issues/29420#issuecomment-810731155. Usually, we only set `isMissing` only for tokens. Doing it for non-terminal nodes requires a small API change. Another way of tackling this is by giving a particular diagnostic. e.g. "Missing annon attachment node" diagnostic.

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
$subject.

Fixes #29730 

## Approach
N/A

## Samples
N/A

## Remarks
N/A

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/test/resources/misc/annotations/annotations_assert_10.json [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    },
+    {
+      "kind": "LIST",
+      "hasDiagnostics": true,
+      "children": [
+        {
+          "kind": "MODULE_VAR_DECL",
+          "hasDiagnostics": true,
+          "diagnostics": [
+            "ERROR_METADATA_NOT_ATTACHED_TO_A_TOP_LEVEL_CONSTRUCT"
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/test/resources/misc/annotations/annotations_assert_10.json:17]: @nadeeshaan this diagnostic is enough to identify the dummy construct attached to the metadata right?

Comment (nadeeshaan) [reply compiler/ballerina-parser/src/test/resources/misc/annotations/annotations_assert_10.json:17]: https://github.com/ballerina-platform/ballerina-lang/issues/29420?notification_referrer_id=MDE4Ok5vdGlmaWNhdGlvblRocmVhZDE3MjA5ODgyMTQ6MTMyOTY3NA%3D%3D&notifications_query=reason%3Amention#issuecomment-811613853

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/test/resources/misc/annotations/annotations_assert_10.json:17]: Resolving as per the https://github.com/ballerina-platform/ballerina-lang/issues/29420#issuecomment-811838312

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -262,7 +262,10 @@ private STNode parseTopLevelNode(STNode metadata) {
         switch (nextToken.kind) {
             case EOF_TOKEN:
                 if (metadata != null) {
-                    addInvalidNodeToNextToken(metadata, DiagnosticErrorCode.ERROR_INVALID_METADATA);
+                    STNode moduleVarDecl = parseModuleVarDecl(metadata);
+                    moduleVarDecl = SyntaxErrors.addDiagnostic(moduleVarDecl, 
+                            DiagnosticErrorCode.ERROR_METADATA_NOT_ATTACHED_TO_A_TOP_LEVEL_CONSTRUCT);
+                    return moduleVarDecl;
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:268]: Should we add this to local var decl as well?

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:268]: sorry I have missed it. thanks. Will add :)

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 3]
```diff
@@ -4370,6 +4370,67 @@ private STNode createModuleVarDeclaration(STNode metadata, STNode publicQualifie
...
+    private STNode createSimpleVarDeclaration(STNode metadata, STNode publicQualifier, List<STNode> qualifiers,
+                                              boolean isModuleVar) {
+        STNode emptyNode = STNodeFactory.createEmptyNode();
+        STNode simpleTypeDescIdentifier = SyntaxErrors.createMissingTokenWithDiagnostics(SyntaxKind.IDENTIFIER_TOKEN,
+                DiagnosticErrorCode.ERROR_MISSING_TYPE_DESC);
+        STNode identifier = SyntaxErrors.createMissingTokenWithDiagnostics(SyntaxKind.IDENTIFIER_TOKEN,
+                DiagnosticErrorCode.ERROR_MISSING_IDENTIFIER);
+        STNode simpleNameRef = STNodeFactory.createSimpleNameReferenceNode(simpleTypeDescIdentifier);
+        STNode semicolon = SyntaxErrors.createMissingTokenWithDiagnostics(SyntaxKind.SEMICOLON_TOKEN,
+                DiagnosticErrorCode.ERROR_MISSING_SEMICOLON_TOKEN);
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:4387]: This part is a duplicate of 4409 to 4416. It simply cannot extract to a method though.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 4]
```diff
@@ -4370,6 +4370,67 @@ private STNode createModuleVarDeclaration(STNode metadata, STNode publicQualifie
                 typedBindingPattern, assign, expr, semicolon);
     }
 
+    private STNode createSimpleVarDeclaration(STNode metadata) {
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:4373]:
```suggestion
    private STNode createMissingSimpleVarDecl(STNode metadata) {
```

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 5]
```diff
@@ -4370,6 +4370,67 @@ private STNode createModuleVarDeclaration(STNode metadata, STNode publicQualifie
...
+        return createSimpleVarDeclaration(metadata, publicQualifier, new ArrayList<>(), true);
+    }
+
+    private STNode createSimpleVarDeclaration(STNode metadata, STNode publicQualifier, List<STNode> qualifiers,
+                                              boolean isModuleVar) {
+        STNode emptyNode = STNodeFactory.createEmptyNode();
+        STNode simpleTypeDescIdentifier = SyntaxErrors.createMissingTokenWithDiagnostics(SyntaxKind.IDENTIFIER_TOKEN,
+                DiagnosticErrorCode.ERROR_MISSING_TYPE_DESC);
+        STNode identifier = SyntaxErrors.createMissingTokenWithDiagnostics(SyntaxKind.IDENTIFIER_TOKEN,
+                DiagnosticErrorCode.ERROR_MISSING_IDENTIFIER);
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:4384]:
```suggestion
                DiagnosticErrorCode.ERROR_MISSING_VARIABLE_NAME);
```

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 6]
```diff
@@ -4370,6 +4370,67 @@ private STNode createModuleVarDeclaration(STNode metadata, STNode publicQualifie
...
+        return STNodeFactory.createVariableDeclarationNode(metadata, emptyNode, typedBindingPattern, emptyNode,
+                emptyNode, semicolon);
+    }
+
+    private STNode createMissingSimpleObjectField(STNode metadata, List<STNode> qualifiers, boolean isObjectTypeDesc) {
+        STNode emptyNode = STNodeFactory.createEmptyNode();
+        STNode simpleTypeDescIdentifier = SyntaxErrors.createMissingTokenWithDiagnostics(SyntaxKind.IDENTIFIER_TOKEN,
+                DiagnosticErrorCode.ERROR_MISSING_TYPE_DESC);
+        STNode identifier = SyntaxErrors.createMissingTokenWithDiagnostics(SyntaxKind.IDENTIFIER_TOKEN,
+                DiagnosticErrorCode.ERROR_MISSING_IDENTIFIER);
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:4413]:
```suggestion
                DiagnosticErrorCode.ERROR_MISSING_VARIABLE_NAME);
```

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: added `ERROR_MISSING_FIELD_NAME` instead as per the offline discussion

**Review by rdulmina — APPROVED** [review]

**Pull request merged**